compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -3111,6 +3124,662 @@ class JSCodeGen()(implicit ctx: Context) {
     }
     if (found == null) None else Some(found)
   }
+
+  /** Gen JS code for a Match, i.e., a switch-able pattern match.
+   *
+   *  In most cases, this straightforwardly translates to a Match in the IR,
+   *  which will eventually become a `switch` in JavaScript.
+   *
+   *  However, sometimes there is a guard in here, despite the fact that
+   *  matches cannot have guards (in the JVM nor in the IR). The JVM backend
+   *  emits a jump to the default clause when a guard is not fulfilled. We
+   *  cannot do that, since we do not have arbitrary jumps. We therefore use
+   *  a funny encoding with two nested `Labeled` blocks. For example,
+   *  {{{
+   *  x match {
+   *    case 1 if y > 0 => a
+   *    case 2          => b
+   *    case _          => c
+   *  }
+   *  }}}
+   *  arrives at the back-end as
+   *  {{{
+   *  x match {
+   *    case 1 =>
+   *      if (y > 0)
+   *        a
+   *      else
+   *        default()
+   *    case 2 =>
+   *      b
+   *    case _ =>
+   *      default() {
+   *        c
+   *      }
+   *  }
+   *  }}}
+   *  which we then translate into the following IR:
+   *  {{{
+   *  matchResult[I]: {
+   *    default[V]: {
+   *      x match {
+   *        case 1 =>
+   *          return(matchResult) if (y > 0)
+   *            a
+   *          else
+   *            return(default) (void 0)
+   *        case 2 =>
+   *          return(matchResult) b
+   *        case _ =>
+   *          ()
+   *      }
+   *    }
+   *    c
+   *  }
+   *  }}}
+   */
+  def genMatch(tree: Tree, isStat: Boolean): js.Tree = {
+    implicit val pos: SourcePosition = tree.sourcePos
+    val Match(selector, cases) = tree
+
+    /* Although GenBCode adapts the scrutinee and the cases to `int`, only
+     * true `int`s can reach the back-end, as asserted by the String-switch
+     * transformation in `cleanup`. Therefore, we do not adapt, preserving
+     * the `string`s and `null`s that come out of the pattern matching in
+     * Scala 2.13.2+.
+     */
+    val genSelector = genExpr(selector)
+
+    val resultType =
+      if (isStat) jstpe.NoType
+      else toIRType(tree.tpe)
+
+    val defaultLabelSym = cases
+      .collectFirst {
+        case CaseDef(Ident(nme.WILDCARD), EmptyTree, body @ Labeled(_, rhs))
```

Comment:
As I said on the issue. You can scrap everything that takes care of `Labeled` things. In scalac we have to reverse-engineer `LabelDef` definitions and jumps to them, but not in Dotty, because the `Labeled` blocks are correctly compiled to the Scala.js IR once and for all in `case Labeled(bind, expr) =>`.

Reply:
Thanks for the explanation, but I'm still too much of a newbie to understand what exactly I can remove. Is there any documentation about this?

Can I remove the whole `defaultLabelSym` value? What shall I do where it is used, e.g.

```
        case app @ Apply(_, Nil) if app.symbol == defaultLabelSym =>
          genJumpToElseClause
```

Reply:
Yes, you can remove it. Basically you can consider that `defaultLabelSym` is always `NoSymbol`. That means that `if app.symbol == defaultLabelSym` is basically `if false`, and that consequently you can remove that `case` entirely.

From there, you'll figure that `genJumpToElseClause` is dead code, and hence `optElseClauseLabel` will never be assigned to a `Some`. You can therefore replace it by `None` and remove dead code from there as well.

The domino effect of this strategy will cause a lot of code to be removed.